src/Microsoft.ML.Transforms/GcnTransform.cs

```diff
@@ -225,477 +255,725 @@ public void Save(ModelSaveContext ctx)
-        private const Float MinScale = (Float)1e-8;
+        private const float MinScale = (float)1e-8;
 
-        private readonly ColInfoEx[] _exes;
+        public IReadOnlyCollection<ColumnInfoBase> Columns => _columns.AsReadOnly();
+        private readonly ColumnInfoBase[] _columns;
 
-        /// <summary>
-        /// A helper method to create GlobalContrastNormalizer transform for public facing API.
-        /// </summary>
-        /// <param name="env">Host Environment.</param>
-        /// <param name="input">Input <see cref="IDataView"/>. This is the output from previous transform or loader.</param>
-        /// <param name="name">Name of the output column.</param>
-        /// <param name="source">Name of the column to be transformed. If this is null '<paramref name="name"/>' will be used.</param>
-        /// <param name="subMean">Subtract mean from each value before normalizing.</param>
-        /// <param name="useStdDev">Normalize by standard deviation rather than L2 norm.</param>
-        /// <param name="scale">Scale features by this value.</param>
-        public static IDataTransform CreateGlobalContrastNormalizer(IHostEnvironment env,
-            IDataView input,
-            string name,
-            string source = null,
-            bool subMean = Defaults.GcnSubMean,
-            bool useStdDev = Defaults.UseStdDev,
-            Float scale = Defaults.Scale)
-        {
-            var args = new GcnArguments()
-            {
-                Column = new[] { new GcnColumn(){
-                        Source = source ?? name,
-                        Name = name
-                    }
-                },
-                SubMean = subMean,
-                UseStdDev = useStdDev,
-                Scale = scale
-            };
-            return new LpNormNormalizerTransform(env, args, input);
+        private static (string input, string output)[] GetColumnPairs(ColumnInfoBase[] columns)
+        {
+            Contracts.CheckValue(columns, nameof(columns));
+            return columns.Select(x => (x.Input, x.Output)).ToArray();
+        }
+
+        protected override void CheckInputColumn(ISchema inputSchema, int col, int srcCol)
+        {
+            var inType = inputSchema.GetColumnType(srcCol);
+            var reason = TestColumn(inType);
+            if (reason != null)
+                throw Host.ExceptParam(nameof(inputSchema), reason);
+        }
+
+        // Check if the input column's type is supported. Note that only float vector with a known shape is allowed.
+        internal static string TestColumn(ColumnType type)
+        {
+            if ((type.IsVector && !type.IsKnownSizeVector && (type.AsVector.Dimensions.Length > 1)) || type.ItemType != NumberType.R4)
+                return "Expected float or float vector of known size";
+
+            if ((long)type.ValueCount * type.ValueCount > Utils.ArrayMaxSize)
+                return "Vector size exceeds limit";
```

Comment:
>Vector size exceeds limit [](start = 24, length = 25)

This probably cannot be put into `ExceptSchemaMismatch`, so make it `ExceptParam`. Also, it would be helpful to spell out what the `limit` is equal to

Reply:
I mean, don't use `TestType`, turn it into `void CheckType` and throw exceptions. I don't like the idea of a method that produces a string error message

---
In reply to: [232437056](https://github.com/dotnet/machinelearning/pull/1582#discussion_r232437056) [](ancestors = 232437056)